docs/architecture/adr-046-module-params.md

```diff
@@ -0,0 +1,165 @@
+field along with a handler. This can become burdensome if there are a lot of
+parameter fields. In the latter case, there is only a single data structure and
+thus only a single message handler, however, the message handler might have to be
+more sophisticated in that it might need to understand what parameters are being
+changed vs what parameters are untouched.
+
+Once the `sdk.Msg` types and corresponding handlers are defined, proposals can
+be made using the `x/gov` module that are authorized by the root `x/gov`
+module's account.
+
+Continuing to use `x/auth`, we demonstrate a more complete example:
+
+```go
+type Params struct {
+	MaxMemoCharacters      uint64
+	TxSigLimit             uint64
+	TxSizeCostPerByte      uint64
+	SigVerifyCostED25519   uint64
+	SigVerifyCostSecp256k1 uint64
+}
+
+type MsgUpdateParams struct {
+	MaxMemoCharacters      uint64
+	TxSigLimit             uint64
+	TxSizeCostPerByte      uint64
+	SigVerifyCostED25519   uint64
+	SigVerifyCostSecp256k1 uint64
+}
+
+type MsgUpdateParamsResponse struct {}
+
+func (ms msgServer) UpdateParams(goCtx context.Context, msg *types.MsgUpdateParams) (*types.MsgUpdateParamsResponse, error) {
+  ctx := sdk.UnwrapSDKContext(goCtx)
+
+  // verification logic...
+
+  // persist params
+  params := ParamsFromMsg(msg)
+  ms.SaveParams(ctx, params)
+
+  return &types.MsgUpdateParamsResponse{}, nil
+}
+
+func ParamsFromMsg(msg *types.MsgUpdateParams) Params {
+  // ...
+}
+```
+
+## Consequences
+
+As a result of implementing the module parameter methodology, we gain the ability
+for module parameter changes to be stateful and extensible to fit nearly every
+application's use case. In addition, there will be significant gains in performance
+when it comes to reading and writing parameters from and to state, especially if
+a specific set of parameters are read on a consistent basis.
+
+However, this methodology will require developers to implement more types and
+handlers which can become burdensome if many parameters exist. In addition,
+developers are required to implement persistance logics of module parameters.
+However, this should be trivial.
+
+### Backwards Compatibility
+
+The new method for working with module parameters is naturally not backwards
+compatible with the existing `x/params` module. However, the `x/params` will
+remain in the Cosmos SDK and will be marked as deprecated with no additional
+functionality being added apart from potential bug fixes. Note, the `x/params`
+module may be removed entirely in a future release.
+
+### Positive
+
+- Module parameters are serialized more efficiently
+- Module parameters changes are able to be stateful
+
+### Negative
+
+- Module parameter UX becomes slightly more burdensome for developers:
+    - Modules are now responsible for persisting and retrieving parameter state
+    - Modules are now required to have unique message handlers to handle parameter
+      changes per unique parameter data structure.
```

Comment:
So this effectively makes params no different than any other data in state: module devs manually define their their queries and state transitions.

Since params are such a common pattern, i wonder if there's a way to factorize this pattern, without going all the way into creating a `x/params` module.

Reply:
> So this effectively makes params no different than any other data in state: module devs manually define their their queries and state transitions.

Yes, correct. The main distinction is they're treated as an `sdk.Msg`.